tchannel/tracing.py

```diff
@@ -0,0 +1,206 @@
+
+from opentracing.ext import tags
+from tchannel.messages import common
+
+
+log = logging.getLogger('tchannel')
+
+
+# noinspection PyMethodMayBeStatic
+class TracingContextProvider(object):
+    """Tracks the OpenTracing Span currently in effect.
+
+    Distributed tracing requires context propagation inside the application,
+    do that outbound (downstream) calls can be part of the same trace as the
+    inbound request.
+
+    There are two ways to do this:
+
+    * Pass some "context" object to every method that may need to use it.
+      This has the best performance but is not always practical.
+
+    * Store the "context" in a thread-local storage, so that it's always
+      accessible from anywhere in the application.
+
+    The asynchronous nature of Tornado means that multiple requests can be
+    in-flight at any given moment, even in the same thread, so plain thread
+    local storage does not work. This class provide an alternative to thread
+    local storage by using Tornado's ``StackContext`` functionality.
+
+    There's currently no way to disable Span tracking via ``StackContext``.
+    """
+    def get_current_span(self):
+        """
+        :return: The current :py:class:`Span` for this thread/request.
+        """
+        return opentracing_instrumentation.get_current_span()
+
+    def span_in_context(self, span):
+        """
+        Store the `span` in the request context and return a `StackContext`.
+
+        This method is meant to be used as a context manager:
+
+        .. code-block:: python
+
+            with tchannel.context_provider.span_in_context(span):
+                f = handler_fn()
+            res = yield f
+
+        Note: StackContext does not allow yield when used a context manager.
+        Instead, save the future and yield it outside of `with:` statement.
+
+        :param span: an OpenTracing Span
+        :return: ``StackContext``-based context manager
+        """
+        return opentracing_instrumentation.span_in_stack_context(span)
+
+
+class ServerTracer(object):
+    """Helper class for creating server-side spans."""
+
+    __slots__ = ['tracer', 'operation_name', 'span']
+
+    def __init__(self, tracer, operation_name):
+        self.tracer = tracer
+        self.operation_name = operation_name
+        self.span = None
+
+    # noinspection PyMethodMayBeStatic
+    def start_basic_span(self, tracing):
+        """
+        Start tracing span from the protocol's `tracing` fields.
+
+        This will only work if the `tracer` supports Zipkin-style span context.
+
+        :param tracing: common.Tracing
+        """
+        pass  # TODO(ys) implement start_basic_span
+
+    def start_span(self, request, headers, peer_host, peer_port):
+        """
+        Starts a new server-side span. If the span has already been started
+        by `start_basic_span`, this method only adds baggage from the headers.
+        :param request:
+        :param headers:
+        :return:
```

Comment:
Please document the args and the return type. This looks like `request` must always be `tchannel.tornado.Request` since `request.headers` is assumed to be the transport headers.